File: Source/Utility/Hooks/use-input-service.hook.ts
import { useEventListener } from "@rbxts/pretty-roact-hooks";
import { UserInputService } from "@rbxts/services";
import { useState } from "@rbxts/roact-hooked";

export type InputDevice = "Keyboard" | "Gamepad" | "Touch";

const getInputType = (Input = UserInputService.GetLastInputType()): InputDevice | undefined => {
	if ((Input === Enum.UserInputType.Keyboard) || (Input === Enum.UserInputType.MouseMovement)) {
		return "Keyboard";

	} else if (Input === Enum.UserInputType.Gamepad1) {
		return "Gamepad";

	} else if (Input === Enum.UserInputType.Touch) {
		return "Touch";
	};
};

/**
 * Returns the current input device being used by the player.
 * @returns An InputDevice string.
 */
export function useInputDevice() {
	const [ Device, SetDevice ] = useState<InputDevice>(() => (getInputType() ?? "Keyboard"));

	useEventListener(UserInputService.LastInputTypeChanged, Input => {
		const NewDevice = getInputType(Input);

		if (NewDevice !== undefined) SetDevice(NewDevice);
	});

	return Device;
};

File: Output/Utility/Hooks/use-input-service.hook.lua
-- Compiled with roblox-ts v2.2.0
local TS = _G[script]
local useEventListener = TS.import(script, TS.getModule(script, "@rbxts", "pretty-roact-hooks").out).useEventListener
local UserInputService = TS.import(script, TS.getModule(script, "@rbxts", "services")).UserInputService
local useState = TS.import(script, TS.getModule(script, "@rbxts", "roact-hooked").src).useState
local getInputType = function(Input)
	if Input == nil then
		Input = UserInputService:GetLastInputType()
	end
	if (Input == Enum.UserInputType.Keyboard) or (Input == Enum.UserInputType.MouseMovement) then
		return "Keyboard"
	elseif Input == Enum.UserInputType.Gamepad1 then
		return "Gamepad"
	elseif Input == Enum.UserInputType.Touch then
		return "Touch"
	end
end
--[[
	*
	 * Returns the current input device being used by the player.
	 * @returns An InputDevice string.
	 
]]
local function useInputDevice()
	local Device, SetDevice = useState(function()
		return getInputType() or "Keyboard"
	end)
	useEventListener(UserInputService.LastInputTypeChanged, function(Input)
		local NewDevice = getInputType(Input)
		if NewDevice ~= nil then
			SetDevice(NewDevice)
		end
	end)
	return Device
end
return {
	useInputDevice = useInputDevice,
}
